### ISSUE TITLE
TP2 hang with HiCache, breakable prefill CUDA graphs, and FlashInfer MNNVL on B300

### ISSUE THREAD

**Issue by nvpohanh** [sgl-project/sglang#38300]
[by Codex]

### Environment

- 2 x NVIDIA B300 in one NVLink domain
- `lmsysorg/sglang:v0.5.18-cu130`
- SGLang v0.5.18, FlashInfer 0.6.17

### Reproduction

1. Download only the public model configuration/tokenizer files and make a 24-layer dummy-weight configuration:

```bash
hf download nvidia/GLM-5.2-NVFP4 \
  --exclude '*.safetensors' \
  --local-dir /tmp/GLM-5.2-NVFP4-dummy

python3 -c 'import json; p="/tmp/GLM-5.2-NVFP4-dummy/config.json"; d=json.load(open(p)); d["num_hidden_layers"]=24; [d.__setitem__(k,d[k][:24]) for k in ("layer_types","mlp_layer_types","indexer_types") if k in d]; json.dump(d,open(p,"w"),indent=2)'
```

2. In the SGLang container, start the server:

```bash
export MC_FORCE_MNNVL=1
export NCCL_MNNVL_ENABLE=1
export NCCL_CUMEM_ENABLE=1

python3 -m sglang.launch_server \
  --model-path /tmp/GLM-5.2-NVFP4-dummy --load-format dummy \
  --tp-size 2 --mem-fraction-static 0.85 \
  --max-total-tokens 250000 --context-length 1048576 \
  --quantization modelopt_fp4 --kv-cache-dtype fp8_e4m3 \
  --attention-backend nsa --nsa-prefill-backend trtllm --nsa-decode-backend trtllm \
  --moe-runner-backend flashinfer_trtllm \
  --flashinfer-allreduce-fusion-backend mnnvl \
  --chunked-prefill-size 32768 --max-running-requests 32 \
  --cuda-graph-config '{"prefill":{"backend":"breakable","max_bs":32768,"bs":[4,8,16,32,64,128,256,512,1024,2048,4096,8192,16384,26752,32768]}}' \
  --speculative-algorithm EAGLE --speculative-num-steps 5 \
  --speculative-eagle-topk 1 --speculative-num-draft-tokens 6 \
  --page-size 64 --enable-hierarchical-cache --hicache-size 160 \
  --enable-cache-report --enable-metrics
```

3. Save the example below as `hicache_hang_load.py` on the same host. It performs the cache-fill phase, verifies that eviction has started, keeps exactly 96 draw requests in flight, runs the CUDA-context churn workers on GPUs 0 and 1, and reports a reproduction after 120 seconds without a completion.

4. Run it while the server is healthy:

```bash
python3 hicache_hang_load.py --url http://127.0.0.1:30000
```

### Observed outcome

Within approximately 1-6 minutes, request progress permanently stops while both scheduler processes remain alive. One GPU becomes idle and the other stays at 100% utilization in a collective kernel. The busy GPU is typically stuck in FlashInfer's MNNVL `twoshotAllreduceKernel`; either TP rank can be the busy rank.

### Expected outcome

All requests complete without a permanent TP-rank hang.

<details>
<summary>Complete load and CUDA-context-churn script</summary>

```python
#!/usr/bin/env python3
"""Load generator for sgl-project/sglang#38300.

Run this on the same two-GPU host as the SGLang server. The Python environment
needs aiohttp and a CUDA-enabled PyTorch installation.
"""

import argparse
import asyncio
import os
import subprocess
import sys
import threading
import time
import urllib.request

import aiohttp


CHURN_CODE = r"""
import torch

a = torch.randn((1024, 1024), device="cuda", dtype=torch.bfloat16)
b = torch.randn((1024, 1024), device="cuda", dtype=torch.bfloat16)
while True:
    c = a @ b
"""


def http_get(url: str, timeout: float = 5) -> str:
    with urllib.request.urlopen(url, timeout=timeout) as response:
        return response.read().decode()


def wait_ready(base_url: str, timeout: float = 1200) -> None:
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        try:
            http_get(f"{base_url}/health", timeout=2)
            return
        except Exception:
            time.sleep(2)
    raise TimeoutError(f"server did not become ready within {timeout}s")


def eviction_seconds(base_url: str) -> float:
    metrics = http_get(f"{base_url}/metrics")
    for line in metrics.splitlines():
        if line.startswith("sglang:eviction_duration_seconds_sum"):
            return float(line.rsplit(" ", 1)[1])
    return 0.0


async def request(
    session: aiohttp.ClientSession,
    base_url: str,
    request_id: int,
    prompt_tokens: int,
) -> None:
    # The first token makes every prefix unique. Both token IDs are within the
    # public GLM-5.2 vocabulary.
    input_ids = [1000 + (request_id % 120000)] + [42] * (prompt_tokens - 1)
    payload = {
        "input_ids": input_ids,
        "sampling_params": {"temperature": 0, "max_new_tokens": 4},
    }
    async with session.post(f"{base_url}/generate", json=payload) as response:
        body = await response.text()
        if response.status != 200:
            raise RuntimeError(f"HTTP {response.status}: {body[:300]}")


async def run_fixed_batch(
    base_url: str,
    first_request_id: int,
    lengths: list[int],
    concurrency: int,
) -> None:
    timeout = aiohttp.ClientTimeout(total=900)
    connector = aiohttp.TCPConnector(limit=concurrency)
    semaphore = asyncio.Semaphore(concurrency)
    async with aiohttp.ClientSession(timeout=timeout, connector=connector) as session:

        async def limited(offset: int, length: int) -> None:
            async with semaphore:
                await request(session, base_url, first_request_id + offset, length)

        await asyncio.gather(
            *(limited(offset, length) for offset, length in enumerate(lengths))
        )


def churn_worker(stop: threading.Event, gpu: int) -> None:
    env = os.environ.copy()
    env["CUDA_VISIBLE_DEVICES"] = str(gpu)
    while not stop.is_set():
        process = subprocess.Popen(
            [sys.executable, "-c", CHURN_CODE],
            env=env,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        stop.wait(2.2)
        if process.poll() is None:
            process.kill()
        try:
            process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            process.kill()
            process.wait()
        stop.wait(6.3)


async def draw_phase(
    base_url: str,
    gpu_ids: list[int],
    duration: float,
    stall_timeout: float,
) -> bool:
    stop = threading.Event()
    churn_threads = [
        threading.Thread(target=churn_worker, args=(stop, gpu), daemon=True)
        for gpu in gpu_ids
    ]
    for thread in churn_threads:
        thread.start()

    timeout = aiohttp.ClientTimeout(total=900)
    connector = aiohttp.TCPConnector(limit=96)
    session = aiohttp.ClientSession(timeout=timeout, connector=connector)
    lengths = (2048, 4096, 26752, 32768)
    pending: set[asyncio.Task] = set()
    started = last_completion = time.monotonic()
    launched = completed = 0
    next_report = started

    try:
        while time.monotonic() - started < duration:
            while len(pending) < 96:
                request_id = 1_000_000 + launched
                length = lengths[launched % len(lengths)]
                pending.add(
                    asyncio.create_task(
                        request(session, base_url, request_id, length)
                    )
                )
                launched += 1

            done, pending = await asyncio.wait(
                pending, timeout=1, return_when=asyncio.FIRST_COMPLETED
            )
            for task in done:
                task.result()
                completed += 1
                last_completion = time.monotonic()

            now = time.monotonic()
            if now >= next_report:
                print(
                    f"elapsed={now-started:.0f}s launched={launched} "
                    f"completed={completed} stalled={now-last_completion:.0f}s "
                    f"eviction={eviction_seconds(base_url)}",
                    flush=True,
                )
                next_report = now + 10

            if pending and now - last_completion >= stall_timeout:
                print(
                    f"HANG_REPRODUCED: no request completed for "
                    f"{now-last_completion:.0f}s",
                    flush=True,
                )
                return True

        print(f"NO_HANG: completed {completed} requests in {duration:.0f}s")
        return False
    finally:
        stop.set()
        for task in pending:
            task.cancel()
        await session.close()
        for thread in churn_threads:
            thread.join(timeout=10)


async def main(args: argparse.Namespace) -> bool:
    wait_ready(args.url)

    # One round is exactly 96 unique prefixes: five 4096-token prompts followed
    # by one 26752-token prompt, repeated 16 times.
    fill_lengths = ([4096] * 5 + [26752]) * 16
    next_request_id = 0
    for round_number in range(1, args.max_fill_rounds + 1):
        await run_fixed_batch(args.url, next_request_id, fill_lengths, concurrency=96)
        next_request_id += len(fill_lengths)
        eviction = eviction_seconds(args.url)
        print(
            f"fill_round={round_number} requests={next_request_id} "
            f"eviction={eviction}",
            flush=True,
        )
        if eviction > 0:
            break
    else:
        raise RuntimeError(
            "HiCache eviction did not start; increase --max-fill-rounds or check "
            "the server's HiCache configuration"
        )

    return await draw_phase(
        args.url, args.gpus, args.duration, args.stall_timeout
    )


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--url", default="http://127.0.0.1:30000")
    parser.add_argument("--gpus", type=int, nargs="+", default=[0, 1])
    parser.add_argument("--duration", type=float, default=720)
    parser.add_argument("--stall-timeout", type=float, default=120)
    parser.add_argument("--max-fill-rounds", type=int, default=4)
    arguments = parser.parse_args()
    reproduced = asyncio.run(main(arguments))
    raise SystemExit(0 if reproduced else 2)
```

</details>

**Comment by nvpohanh** [comment]
[by Codex]

Additional debugging details:

- Both TP ranks choose the same FlashInfer AllReduce tactics. The captured CUDA graphs are identical between ranks, as are their graph replay sequences and uncaptured FlashInfer dispatches.
- The final failing replay is a breakable prefill graph for batch size 512 containing 49 MNNVL two-shot nodes. Both ranks submit the same graph and all 33 of its segments.
- Device-side journals written after `cudaGridDependencySynchronize()` show both ranks entering the first two-shot in segment 2. One rank then enters the second two-shot in that segment while the other never reaches its corresponding entry marker.
- The ahead rank is active in `twoshotAllreduceKernel<2, __nv_bfloat16, false, float4>`, grid `(512,6,1)`, block `(128,1,1)`. CUDA-GDB places it in `reduceLamportRanksChunked<2,2,...>`, polling the peer's Lamport scatter slot. The other rank has an active CUDA context but no active CUDA kernel.
- The direct graph neighborhood is an NVJet GEMM, followed by the MNNVL two-shot kernel, followed by FlashInfer `rmsNormLamport`. It is identical on both ranks.
- `GenerationBatchResult.copy_done` is created and recorded correctly on both ranks. A diagnostic event recorded directly on `forward_stream` also remains incomplete, proving that `copy_done` is downstream of the actual stall rather than its cause.
- The negative-zero/subnormal hardening from FlashInfer PR #3304 is present in this MNNVL two-shot path. The observed peer contribution is genuinely missing because the peer never reaches the matching kernel; this is not a flush-to-zero sentinel misclassification.
- Disabling overlap scheduling does not eliminate the broader workload hang, although that control stopped in NCCL rather than MNNVL and therefore is not a clean MNNVL-specific control.
- Fully disabling breakable prefill CUDA graphs avoids the hang in the controls run so far. `--disable-cuda-graph` alone is insufficient when an explicit breakable prefill graph configuration is supplied; prefill graphs must also be disabled.

The current unresolved boundary is execution progress between the first and second MNNVL nodes in segment 2. The next instrumentation will distinguish whether the lagging rank never launches the second node or launches it but remains before/inside `cudaGridDependencySynchronize()`, and will add source call-site/layer tags around the intervening graph nodes.

**Comment by mattteochen** [comment]
AI assisted investiagtion.

I reproduced this and bisected it down: the hang only happens when another process's CUDA context gets torn down on the serving GPU while HiCache (default `kernel` IO backend) and a CUDA-graph replay are active, and at that moment the idle rank's GPU just stops dispatching the in-flight graph with no kernel running and no error, so the busy rank is only waiting for it in the next all-reduce; this points at a driver/GPU context-teardown issue rather than a collective or SGLang bug. The good news is that `--hicache-io-backend direct` avoided the hang in three soaks including your exact script, while disabling prefill CUDA graphs alone did not.

**Comment by nvpohanh** [comment]
[by Codex]

Debugging update: the strongest current root-cause candidate is a missing stream dependency after HiCache device-to-host writes.

With overlap scheduling, the scheduler loop runs on `schedule_stream`. `L2TransferEngine.submit_device_to_host()` records its dependency event on that stream and makes the private D-to-H stream wait for it. Later, `run_batch()` makes `forward_stream` wait for `schedule_stream`, but neither stream waits for the D-to-H finish event. The effective dependency graph is therefore a fork:

```text
schedule_stream --start_event--> HiCache D-to-H stream
       |
       +------------------------> forward_stream / CUDA graph
```

This allows an unfinished HiCache KV write to overlap a later graph replay. Under the reproducer's secondary CUDA-context/GEMM pressure, one TP rank can stop before its next collective while its peer enters MNNVL/NCCL and spins indefinitely.

Several graph-enabled isolation experiments support this:

- Recycling all per-write CUDA events from a bounded pool did not fix the hang; it reproduced after about 507 seconds.
- Replacing the paged allocator's `torch.unique` free path with fixed-shape `free_segment()` removed substantial synchronization overhead but did not fix the hang; it reproduced after about 115 seconds.
- Synchronizing every D-to-H completion at submission prevented the hang for a 720-second run.
- More importantly, adding an asynchronous wait for the existing completion event on the current `schedule_stream` also prevented the hang:

```python
completion = self.l2_transfer_engine.submit_device_to_host(...)
completion.finish_event.wait()
```

Because `forward_stream` already waits for `schedule_stream`, this closes the fork without blocking the host:

```text
schedule --start--> D-to-H --finish--> schedule --> forward graph
```

The fence-only variant was tested in a fresh container with the original event lifecycle and original allocator free path. It completed 720 seconds and 2,927 requests without a hang when GPU 0 was subjected to the secondary-context workload, and independently completed 720 seconds and 2,927 requests when that workload was moved to GPU 1. Full CUDA graphs for decode and breakable piecewise CUDA graphs for prefill remained enabled.

Our current hypothesis is therefore that the relevant HiCache-induced churn is asynchronous D-to-H KV-transfer traffic on a detached auxiliary stream, and that the missing D-to-H-finish edge before subsequent graph replay is the SGLang-level ordering defect. The one-line wait is a correctness-first fix candidate, not yet a final performance-validated fix: additional soak repetitions and a stock-versus-fenced throughput/latency comparison are still planned.
